### ISSUE TITLE
60 FPS lock in Max Payne

### ISSUE THREAD

**Issue by deton24** [crosire/d3d8to9#95]
I use Widescreen by 13AG as dinput.dll and only d3d8.dll without even d3d9.dll from reshade.
I get 60 FPS lock.
When I delete d3d8.dll (d3d8to9) it unlocks FPS.

AMD card.

**Comment by Sethy1124** [comment]
If your monitor is only 60 Hz then it really doesn't matter as the game will only be 60fps even if an fps counter shows higher

**Comment by HerMajestyDrMona** [comment]
I think adding:
Output.PresentationInterval = D3DPRESENT_INTERVAL_IMMEDIATE;

At the end of function ConvertPresentParameters() in d3d8types.cpp removes the 60FPS limit.

**Comment by elishacloud** [comment]
I checked the game.  Both Max Payne and Max Payne 2 send the following parameters:

* Windowed = false
* SwapEffect = D3DSWAPEFFECT_DISCARD
* FullScreen_RefreshRateInHz = D3DPRESENT_RATE_DEFAULT
* FullScreen_PresentationInterval = D3DPRESENT_INTERVAL_ONE

From everything I have read and looking at the Microsoft documentation [here](https://docs.microsoft.com/en-us/previous-versions/ms886318%28v%3dmsdn.10%29) and [here](https://docs.microsoft.com/en-us/previous-versions/ms886364%28v%3dmsdn.10%29) these settings should **enable** vsync, as d3d8to9 does.

Here are two quotes from Microsoft's documentation:

> D3DSWAPEFFECT_DISCARD
> For a full-screen swap chain, the presentation rate is determined by the value assigned to the `FullScreen_PresentationInterval` member of the `D3DPRESENT_PARAMETERS` structure when the device or swap chain is created. Unless this value is `D3DPRESENT_INTERVAL_IMMEDIATE`, the presentation will be synchronized with the vertical sync of the monitor.

> Flag | Description
> -- | -- 
> D3DPRESENT_INTERVAL_ONE | The driver will wait for the vertical retrace period. Present  operations will not be affected more frequently than the screen refresh.

Either Microsoft's documentation is wrong, or there is a bug in the d3d8 implementation.  I suspect the latter.

I don't believe there is any d3d8to9 bug here.  I recommend either using your graphics card to disable vsync, using some third party tool that can do that or creating a custom build, as HerMajestyDrMona mentioned [above](https://github.com/crosire/d3d8to9/issues/95#issuecomment-462178962).

![image](https://user-images.githubusercontent.com/25625456/73601029-bbbff680-450e-11ea-8049-7f7ca478ea33.png)

**Comment by mirh** [comment]
https://github.com/crosire/d3d8to9/pull/143
TL;DR microsoft removed exclusive fullscreen from d3d8. 
And whether this had always been the case or yet another DWM quirk, thus you get windowed which means unlimited framerate.

**Comment by deton24** [comment]
To raise 60 FPS limit in pre-1809 W10 builds, and while not using windowed mode or custom d3d8to9 build, use RivaTuner Statistics Server, add MaxPayne.exe and set "Custom Direct3D support". It will start to use 120 FPS limit from now on. 
When the option is set when the game is minimized, you must sometimes task switch twice or restart the game to make it work.

Physics issues in Max Payne start somewhere above 144 FPS. Curiously, it ignores whatever I set in Framerate limit. The same goes for whatever is set in Radeon Chill. You can also unlock FPS completely, by setting Radeon Enhanced Sync, but it might cause stuttering if FPS fluctuates too much (and mentioned physics issues), plus if you push OpenALSoft settings for DSOAL too much, the sound might occasionally stutter.


Latest 1.12.0 d3d8to9 used, with also Reshade 2.0 as d3d9.dll, although the app is rather still falsely reported as d3d8 in RivaTuner's OSD.

**Comment by elishacloud** [comment]
@deton24, you can also use [dxwrapper](https://github.com/elishacloud/dxwrapper) with the below settings.  This will force vsync to be disabled and you can set the FPS limit speed.

```ini
EnableVSync                = 0
ForceVsyncMode             = 1
LimitPerFrameFPS           = 119.88
```

**Comment by deton24** [comment]
That will probably have some performance hit (probably not noticeable in
that scenario and with that FPS limit), but it might be useful for a
feature. Thanks!

śr., 18 wrz 2024 o 18:22 Elisha Riedlinger ***@***.***>
napisał(a):
